### ISSUE TITLE
Remove defaultScheduler, observe, drain, reduce

### ISSUE THREAD

**Issue by briancavalier** [mostjs/core#5]
**tl;dr** The defaultScheduler is convenient.  However, it introduces a few things that are less than ideal in my mind.  I propose we remove `defaultScheduler`, `observe`, `drain`, and `reduce` in favor of a single `runStream` primitive.

### Module-load-time side effect

The creation of defaultScheduler is a module-load-time side effect.  This is a minor issue, but not great from a purity standpoint.  If Scheduler (or any of its constituents) is modified to have a non-harmless side-effect, it would automatically impact any program that ends up loading the defaultScheduler module.

### Encourages problematic design patterns

This one is probably much more controversial 😄 

In the Promise world, it is well-established that a good default model is to write functions that accept non-promise parameters, and return a Promise.  Viewing a program as a tree, this encourages propagating promises "up" from program leaf nodes toward the root.  In a lazy Future/Task world, this model is even more desirable because it *centralizes the activation of side effects* at or near the root.  Scattering the activation of side effects out to leaf nodes makes programs much harder to reason about, and any coordination among those scattered side effects will be implicit or ad hoc and may only "work" by happy accident.

Event streams are a similar construct, and imho, it's desirable to program in the same way: to return event streams and centralize observation, which is the event stream version of "activate side effects", at or near the root.

Through time and examples, the "return promises/futures/tasks" model has become an accepted best practice.  Another way to help that process along is to make it unergonomic, or just plain annoying, to observe at leaf nodes.  One way to do *that* is to require that any function that activates an event stream accept a Scheduler instance as a parameter.  So, a program that observes at leaf nodes would need to either: 1) construct lots of Schedulers, which is, imho, fairly obviously The Wrong Thing, and will make it impossible to coordinate side effects among observations that use different Schedulers, or 2) pass a Scheduler down to leaf nodes, which (hopefully) will be quite annoying and ugly.

## Proposal

To encourage the practice of returning event streams and centralizing observation, I propose we:

1. remove defaultScheduler
2. remove observe, drain, and reduce
3. export a function which can be used to build observe, drain, and reduce in the cujojs/most layer
4. export Scheduler and its constituents (Timeline, etc.)
5. export a `createDefaultScheduler()` function (see below)

```hs
type Observer a = { event :: a -> (), error :: Error e => e -> (), end :: () -> () }

runStream :: Observer a -> Scheduler -> Stream a

createDefaultScheduler :: () -> Scheduler
```

**Comment by TylorS** [comment]
I'm Tylor Steinberger, and I approve of this message. :smile:

**Comment by briancavalier** [comment]
See #8

**Comment by briancavalier** [comment]
This is done from a public API and a build perspective: However, defaultScheduler, observe, drain, and reduce all still exist in the codebase, as they are used in unit tests, and it will be a larger effort to remove them.

Closing this issue and will open another for refactoring the unit tests.
